File: 61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # if head is None or head.next is None:
        #     return head
        # count = 0
        # p=head
        # while p != None:
        #     count += 1
        #     p=p.next
        # if k> count :
        #     k=k%count
        # for i in range(k):
        #     p=head
        #     q=None
        #     temp=None
        #     while p.next != None:
        #         q=p
        #         p=p.next
        #     temp = p
        #     q.next=None
        #     temp.next=head
        #     head=temp
        # return head

        p = head
        count = 0
        while p:
            p = p.next
            count += 1
        
        if (count ==0):
            return None
        k = k %count
        if (k == 0):
            return head
        
        p = head
        for i in range(k):
            p = p.next
        
        q = head
        while p.next:
            q = q.next
            p = p.next
        temp = q.next
        q.next = None
        p = head
        head = temp
        while temp.next:
            temp = temp.next
        temp.next = p
        return head
        

        

File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        int count = 0;
        ListNode p = head;
        while (p != null){
            count++;
            p = p.next;
        }
        if(count == 0){
            return null;
        }
        k = k%count;
        p = head;

        if(k == 0){
            return head;
        }

        for(int i =0; i< k;i++){
            p = p.next;
        }
        ListNode q = head;
        while(p.next != null){
            q = q.next;
            p = p.next;
        }
        ListNode temp = q.next;
        q.next = null;
        
        p = head;
        head = temp;
        while(temp.next != null){
            temp = temp.next;
        }
        temp.next = p;
        return head;

    }
}